core/ir/riscv64/ir_utils.c

```diff
@@ -48,14 +55,145 @@ convert_to_near_rel_arch(dcontext_t *dcontext, instrlist_t *ilist, instr_t *inst
+        PRE(ilist, instr, instr_addiw);
+        if (first != NULL && !*first_set) {
+            *first = instr_addiw;
+            *first_set = true;
+        }
+        if (last != NULL)
+            *last = instr_addiw;
+    }
+}
+
+static void
+mov64(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr, opnd_t dst, ptr_int_t val,
+      OUT instr_t **first, OUT instr_t **last, OUT bool *first_set)
+{
+    instr_t *tmp;
+    if (((val << 32) >> 32) == val) {
+        mov32(dcontext, ilist, instr, dst, val, first, last, first_set);
+        return;
+    }
+
+    /*
+     * For 64-bit val, a sequence of up to 8 instructions (i.e.,
+     * LUI+ADDIW+SLLI+ADDI+SLLI+ADDI+SLLI+ADDI) is emitted.
+     *
+     * In the following, val is processed from LSB to MSB while instruction emission is
+     * performed from MSB to LSB by calling mov64 recursively. In each recursion,
+     * the lowest 12 bits are removed from val and the optimal shift amount is calculated.
+     * Then, the remaining part of val is processed recursively and mov32 get called as
+     * soon as it fits into 32 bits.
+     */
+    int64_t lo12 = (val << 52) >> 52;
+    /* Add 0x800 to cancel out the signed extention of ADDI. */
+    int64_t hi52 = (val + 0x800) >> 12;
+    int shift = 12 + trailing_zeros_64((uint64)hi52);
+    hi52 = ((hi52 >> (shift - 12)) << shift) >> shift;
+
+    mov64(dcontext, ilist, instr, dst, hi52, first, last, first_set);
+    tmp = INSTR_CREATE_slli(
+        dcontext, dst, dst,
+        opnd_add_flags(opnd_create_immed_int(shift, OPSZ_6b), DR_OPND_IMM_PRINT_DECIMAL));
+    PRE(ilist, instr, tmp);
+    if (last != NULL)
+        *last = tmp;
+
+    if (lo12) {
+        tmp = INSTR_CREATE_addi(dcontext, dst, dst,
+                                opnd_add_flags(opnd_create_immed_int(lo12, OPSZ_12b),
+                                               DR_OPND_IMM_PRINT_DECIMAL));
+        PRE(ilist, instr, tmp);
+        if (last != NULL)
+            *last = tmp;
+    }
+}
+
+/* FIXME i#3544: Keep this in sync with patch_mov_immed_arch(), which is not implemented
+ * yet. */
 void
 insert_mov_immed_arch(dcontext_t *dcontext, instr_t *src_inst, byte *encode_estimate,
                       ptr_int_t val, opnd_t dst, instrlist_t *ilist, instr_t *instr,
                       OUT instr_t **first, OUT instr_t **last)
 {
     /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    ASSERT_NOT_IMPLEMENTED(src_inst == NULL && encode_estimate == NULL);
+
+    CLIENT_ASSERT(opnd_is_reg(dst), "RISC-V cannot store an immediate direct to memory");
+
+    if (opnd_get_reg(dst) == DR_REG_X0) {
+        /* Moving a value to the zero register is a no-op. We insert nothing,
+         * so *first and *last are set to NULL. Caller beware!
+         */
+        if (first != NULL)
+            *first = NULL;
+        if (last != NULL)
+            *last = NULL;
+        return;
+    }
+
+    ASSERT((uint)(opnd_get_reg(dst) - DR_REG_X0) < 32);
```

Comment:
`ASSERT(reg_is_pointer_sized(opnd_get_reg(dst)))` instead?  Its docs say it's supposed to be just GPR's: looks like that's not true for RISCV?  Hmm, I thought "GPR" meant integer and not floating-point.

Reply:
> looks like that's not true for RISCV?

This procedure only applies to integer registers. Refer to https://github.com/DynamoRIO/dynamorio/pull/6095#discussion_r1235969768, we're calling both integer and float registers "general" registers. Generally, GPRs only meant integer registers yes, I thought it has different meanings under DR's context. It's not?

Reply:
Checked again, and RISC-V's  `reg_is_gpr()` is actually correct too, it's just the comment of `DR_REG_START_64`, `DR_REG_STOP_64`, `DR_REG_START_32` and `DR_REG_STOP_32` have an extra "general", I've removed them. Also changed this line to: `ASSERT(reg_is_gpr(opnd_get_reg(dst)));`

Reply:
Ah yes that is what I saw: the "general" for those _64 ones.  OK, good.